src/fsharp/FSharp.Core/control.fs

```diff
@@ -1178,14 +1222,29 @@ namespace Microsoft.FSharp.Control
     // Contains helpers that will attach continuation to the given task.
     // Should be invoked as a part of protectedPrimitive(withResync) call
     module TaskHelpers = 
+        // This uses a trick to get the underlying OperationCanceledException
+        let inline getCancelledException (completedTask:Task) (waitWithAwaiter) =
+            let fallback = new TaskCanceledException(completedTask) :> OperationCanceledException
+            // sadly there is no other public api to retrieve it, but to call .GetAwaiter().GetResult().
+            try waitWithAwaiter()
+                // should not happen, but just in case...
+                fallback
+            with
+            | :? OperationCanceledException as o -> o
+            | other ->
+                // shouldn't happen, but just in case...
+                new TaskCanceledException(fallback.Message, other) :> OperationCanceledException
+
         let continueWith (task : Task<'T>, args, useCcontForTaskCancellation) = 
 
             let continuation (completedTask : Task<_>) : unit =
                 args.aux.trampolineHolder.Protect((fun () ->
                     if completedTask.IsCanceled then
+                        let cancelledException =
+                            getCancelledException completedTask (fun () -> completedTask.GetAwaiter().GetResult() |> ignore)
                         if useCcontForTaskCancellation
-                        then args.aux.ccont (new OperationCanceledException(args.aux.token))
```

Comment:
What's happened to the ``args.aux.token`` argument?  Previously we were passing it in but now `getCancelledException` is being used and AFAICS it isn't setting the token

Reply:
I think this is one of the changes where we changed to an `TaskCanceledException` exception if possible (or leave an existing exception in place). The `TaskCanceledException` improves interop when used from C# (ie when this is Started as Task on an higher level) but
 - The `TaskCanceledException` has no parameter to set a token, we associate with the Task instead
 - If we re-use an `OperationCanceledException` we assume it is correct
 - The third case shouldn't actually happen, but we could use an `OperationCanceledException` with the token there I guess (as long as we set the InnerException it should be fine). I'd prefer the `TaskCanceledException` as that's what actually happend (`completedTask.IsCancelled` is `true`)

Reply:
I think I'll play around with changing this and there is probably one of the new tests that will fail. And I'll try to give some examples. 
Because this might as well be a merge artifact (maybe it has been changed in master in the mean time)

Reply:
Ok currently this doesn't have any observable impact, because `useCcontForTaskCancellation` is only `true` when `FSCORE_PORTABLE_NEW` is defined and only for the `Sleep` implementation (so this change would have changed the observable exception a bit in this scenario).
I changed this part by renaming the `useCcontForTaskCancellation` to `useSimpleCcontForTaskCancellationAndLooseException` to make it more clear that setting this to `true` has quite an impact. I also changed the code that `useSimpleCcontForTaskCancellationAndLooseException` has the old behavior (which I think is fine because `Task.Sleep` will probably never throw user-exceptions).